### PR TITLE
[FIX] base: Update non-stored field error logging to include exc details

### PR DESCRIPTION
**Pull request by abz-odoo**
We have many issues with this log, except that with a `logger.error` we
do not have exception information (exc_info).

As mentionned in the documentation above, we cannot use `logger.exception`
in this case because we are not within an exception handler.
However, we can still retrieve the exception information using `sys`
module and the `exc_info()` method.

https://docs.python.org/3/library/logging.html#logging.Logger.exception
```
exception(msg, *args, **kwargs)
  ...
  This method should only be called from an exception handler.
```

https://docs.python.org/3/library/sys.html#sys.exc_info
```
sys.exc_info()
This function returns the old-style representation of the handled exception.
...
If no exception is being handled anywhere on the stack, this function
return a tuple containing three None values.
...
```

### CODE REVIEW

File: odoo/osv/expression.py [hunk 1]
```diff
@@ -710,7 +710,7 @@ def push_result(query, params):
                 # Non-stored field should provide an implementation of search.
                 if not field.search:
                     # field does not support search!
-                    _logger.error("Non-stored field %s cannot be searched.", field)
+                    _logger.exception("Non-stored field %s cannot be searched.", field)
```
Comment (KangOl) [inline odoo/osv/expression.py:713]: That won't work.
`Logger.exception` only works in the context of an exception (in an `except` block).
What you want to use is the `stack_info` argument.
```suggestion
                    _logger.error("Non-stored field %s cannot be searched.", field, stack_info=True)
```

File: odoo/osv/expression.py [hunk 2]
```diff
@@ -710,7 +710,7 @@ def push_result(query, params):
...
                     if _logger.isEnabledFor(logging.DEBUG):
                         _logger.debug(''.join(traceback.format_stack()))
```
Comment (KangOl) [inline odoo/osv/expression.py:715]: This should then be removed.
```suggestion
```

File: odoo/osv/expression.py (hunk 1 above)
Comment (abz-odoo) [reply odoo/osv/expression.py:713]: Hello, thanks for your reply.

However, I have tried it locally and Sentry is able to retrieve the necessary information using exc_info=True

image

(right: `logger.exception`, left: `logger.error`)

Comment (KangOl) [reply odoo/osv/expression.py:713]: And in the console/log file?

Comment (abz-odoo) [reply odoo/osv/expression.py:713]: In the console it doesnt change anything. Just the normal log without exception, like `logger.error`

So what you're suggesting is to have a stacktrace in the console in addition to Sentry, right?

**Comment by abz-odoo** [comment]
> The commit message needs to explain the nuances of the issue, and why this needs to use `sys.exc_info()` rather than `Logger.exception` & co.

Done

File: odoo/osv/expression.py [hunk 3]
```diff
@@ -710,9 +711,7 @@ def push_result(query, params):
                 # Non-stored field should provide an implementation of search.
                 if not field.search:
                     # field does not support search!
-                    _logger.error("Non-stored field %s cannot be searched.", field)
-                    if _logger.isEnabledFor(logging.DEBUG):
-                        _logger.debug(''.join(traceback.format_stack()))
+                    _logger.error("Non-stored field %s cannot be searched.", field, exc_info=sys.exc_info())
```
Comment (KangOl) [inline odoo/osv/expression.py:714]: We are not in an exception handler, so `sys.exc_info()` will return None, or at worst an unrelated exception. There is the `stack_info` argument for the current traceback.
```suggestion
                    _logger.error("Non-stored field %s cannot be searched.", field, stack_info=True)
```

Comment (KangOl) [reply odoo/osv/expression.py:716]: Unfortunately, `sentry` doesn't respect it.
So using `exc_info=True` should be used.

And the `if _logger.isEnabledFor(logging.DEBUG):` block should be kept.

**Pull request closed without merging**